samples/iothub-connect-to-eventhubs/IotHubConnection.cs

```diff
@@ -0,0 +1,355 @@
+        ///
+        /// <remarks>
+        ///   The link is not opened by this method; callers are required to open the link in order to trigger
+        ///   the redirection error.
+        /// </remarks>
+        ///
+        /// <seealso href="https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-amqp-support"/>
+        ///
+        private static async Task<AmqpLink> CreateRedirectLinkAsync(AmqpConnection connection,
+                                                                    Uri serviceEndpoint,
+                                                                    TimeSpan timeout)
+        {
+            var linkPath = $"{ serviceEndpoint.AbsolutePath }/$management";
+            var session = default(AmqpSession);
+
+            try
+            {
+                var sessionSettings = new AmqpSessionSettings { Properties = new Fields() };
+                session = connection.CreateSession(sessionSettings);
+
+                await session.OpenAsync(timeout).ConfigureAwait(false);
+
+                var linkSettings = new AmqpLinkSettings
+                {
+                    Role = true,
+                    TotalLinkCredit = 1,
+                    AutoSendFlow = true,
+                    SettleType = SettleMode.SettleOnSend,
+                    Source = new Source { Address = linkPath },
+                    Target = new Target { Address = serviceEndpoint.AbsoluteUri }
+                };
+
+                var link = new ReceivingAmqpLink(linkSettings);
+                linkSettings.LinkName = $"{ nameof(IotHubConnection) };{ connection.Identifier }:{ session.Identifier }:{ link.Identifier }";
+                link.AttachTo(session);
+
+                return link;
+            }
+            catch
+            {
+                session?.Abort();
+                throw;
+            }
+        }
+        /// <summary>
+        ///   Builds a shared access signature to use for authentication with the IOT Hub
+        ///   service.
+        /// </summary>
+        ///
+        /// <param name="resourceUri">The address of the resource the signature is intended to authenticate for.</param>
+        /// <param name="keyName">Name of the shared access key to base the signature on.</param>
+        /// <param name="keyValue">The value of the shared access key to base the signature on.</param>
+        /// <param name="validityDuration">The amount of time that the signature should be considered valid; after this time a new signature would be needed.</param>
+        ///
+        /// <returns>The shared access signature, encoded and suitable for use as a credential in service communication.</returns>
+        ///
+        /// <seealso href="https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security#security-tokens" />
+        ///
+        private static string BuildSignature(string resourceUri,
```

Comment:
Too bad this is internal: https://github.com/Azure/azure-iot-sdk-csharp/blob/93128d3625d3bc77d4621fcd0c0579513bf43f15/iothub/device/src/HsmAuthentication/SasTokenBuilder.cs

Reply:
Agreed.   Unfortunately, the closest that is offered is the [IoT Hub Developer Guide](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security#security-tokens) detailing how to create the signature.